source/common/network/apple_dns_impl.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include <dns_sd.h>
+
+#include <cstdint>
+#include <string>
+
+#include "envoy/common/platform.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/event/file_event.h"
+#include "envoy/network/dns.h"
+
+#include "common/common/linked_object.h"
+#include "common/common/logger.h"
+#include "common/common/utility.h"
+
+#include "absl/container/node_hash_map.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * Implementation of DnsResolver that uses Apple dns_sd.h APIs. All calls and callbacks are assumed
+ * to happen on the thread that owns the creating dispatcher.
+ */
+class AppleDnsResolverImpl : public DnsResolver, protected Logger::Loggable<Logger::Id::upstream> {
+public:
+  AppleDnsResolverImpl(Event::Dispatcher& dispatcher);
+  ~AppleDnsResolverImpl() override;
+
+  // Network::DnsResolver
+  ActiveDnsQuery* resolve(const std::string& dns_name, DnsLookupFamily dns_lookup_family,
+                          ResolveCb callback) override;
+
+private:
+  struct PendingResolution : public ActiveDnsQuery {
+    PendingResolution(AppleDnsResolverImpl& parent, ResolveCb callback,
+                      Event::Dispatcher& dispatcher, DNSServiceRef sd_ref,
+                      const std::string& dns_name)
+        : parent_(parent), callback_(callback), dispatcher_(dispatcher), individual_sd_ref_(sd_ref),
```

Comment:
It's odd to me that you initialize the individual_sd_ref_ with the main ref. Maybe this is just how the API works. More comments?

Reply:
Yep, that's how the API works. Will add more inline comments. This is definitely not the most intuitive API 😅